File: apps/web/lib/apiKeys.ts
import { api } from "@/lib/api";

export type ApiKeyType = "widget" | "secret";

export interface ApiKeyItem {
  id: string;
  name: string;
  key_type: ApiKeyType;
  prefix: string;
  // Plaintext for widget keys (public by design); null for secret keys.
  public_value: string | null;
  scopes: string[];
  last_used_at: string | null;
  revoked_at: string | null;
  created_at: string;
}

export interface CreatedApiKey extends ApiKeyItem {
  token: string;
}

export function listApiKeys(): Promise<ApiKeyItem[]> {
  return api<ApiKeyItem[]>("/api/v1/keys");
}

export function createApiKey(name: string, keyType: ApiKeyType): Promise<CreatedApiKey> {
  return api<CreatedApiKey>("/api/v1/keys", {
    method: "POST",
    body: JSON.stringify({ name, key_type: keyType }),
  });
}

export function revokeApiKey(id: string): Promise<ApiKeyItem> {
  return api<ApiKeyItem>(`/api/v1/keys/${id}`, { method: "DELETE" });
}

File: apps/api/app/schemas/api_keys.py
import datetime
import uuid

from pydantic import BaseModel, Field

from app.models import ApiKeyType


class ApiKeyResponse(BaseModel):
    id: uuid.UUID
    name: str
    key_type: ApiKeyType
    prefix: str
    # Plaintext for widget keys (public by design); None for secret keys.
    public_value: str | None
    scopes: list[str]
    last_used_at: datetime.datetime | None
    revoked_at: datetime.datetime | None
    created_at: datetime.datetime


class ApiKeyCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    key_type: ApiKeyType = ApiKeyType.secret


class ApiKeyCreatedResponse(ApiKeyResponse):
    """Returned only at creation — ``token`` is never recoverable for secret
    keys afterwards (only its hash is stored)."""

    token: str
